File: Python/puzzle8.py
import sys
import unittest
from heapq import *

def to_1d(n, i, j):
    return i * (n-1) + j

def to_2d(n, k):
    row = k / n
    col = k % n
    return row, col

class Board(object):
    """ Immutable representation of puzzle board
        with dimension n and manhattan heurisitc """

    def __init__(self, tiles):
        self.board = tuple([tuple(row) for row in tiles])
        self.n = len(tiles[0])
        self.manhattan = 0
        for i, row in enumerate(tiles):
            for j, tile in enumerate(row):
                goalrow = (tile - 1) / self.n
                goalcol = (tile - 1) % self.n
                if tile != 0:
                    self.manhattan += abs(i - goalrow) + abs(j - goalcol)

    def __eq__(self, other):
        if self.n != other.n: return False
        same = True
        for i in range(self.n):
            same = same and self.board[i] == other.board[i]
        return same

    def __ne__(self, other):
        return not self.__eq__(other)

    def __repr__(self):
        s = "\n"
        for row in self.board:
            s += str(row) + '\n'
        s += '\nManhattan: ' + str(self.manhattan) + '\n'
        return s

    def neighbors(self):
        n = self.n
        flat = [self.board[i][j] for i in range(n) for j in range(n)]
        idx = flat.index(0)
        z_row, z_col = to_2d(n, idx)
        boards = []
        north = (-1, 1, 0, 0)
        east = (0, 0, 1, -1)
        for i in range(4):
            r = z_row + north[i]
            c = z_col + east[i]
            tiles = []
            if not any([r < 0, r >= n, c <0, c >= n]):
                for m in range(n):
                    tiles_row = []
                    for k in range(n):
                        tiles_row.append(self.board[m][k])
                    tiles.append(tiles_row)
                tiles[z_row][z_col] = self.board[r][c]
                tiles[r][c] = 0
                nb = Board(tiles)
                if nb:
                    boards.append(nb)
        return boards

class Node(object):

    def __init__(self, board, m, previous):
        self.board = board
        self.m = m
        self.previous = previous
        self.priority = self.board.manhattan + m  

    def __lt__(self, other):
        return self.priority < other.priority 

def solve(initial_board):
    initial_node = Node(initial_board, 0, None)
    frontier = []
    heappush(frontier, initial_node)
    while True:
        node = heappop(frontier)
        if node.board.manhattan == 0:
            return node
        for b in node.board.neighbors():
            if node.previous == None or b != node.previous.board:
                heappush(frontier, Node(b, node.m+1, node))
    return node

def read_board(f):
    file = open(f)
    tiles = []
    for i, line in enumerate(file):
        if i == 0: continue
        row = [int(x) for x in line.split()]
        tiles.append(row)
    return tiles

def main():
    f = sys.argv[1]
    tiles = read_board(f)
    initial_board = Board(tiles)
    node = solve(initial_board)
    print node.board
    node = node.previous
    while node:
        print node.board
        node = node.previous


class TestSequenceFunctions(unittest.TestCase):

    def setUp(self):
        self.a = Board([[1,2],[3,0]])
        self.b = Board([[2,1,3],[4,5,6],[7,8,0]])
        self.c = Board([[2,1,3],[4,5,6],[7,8,0]])
        self.d = Board([[0,1,3],[4,2,5],[7,8,6]])
        self.e = Board([[1,2,3],[4,5,6],[7,8,0]])


    def test_equals(self):
        self.assertTrue(self.b == self.c)
        self.assertTrue(self.b != self.a)
        self.assertTrue(self.d != self.c)

    def test_neighbors(self):
        self.assertTrue(len(self.d.neighbors()) == 2)
        self.assertTrue(len(self.b.neighbors()) == 2)

    def test_to_1d(self):
        self.assertTrue(to_1d(5,0,0) == 0)
        self.assertTrue(to_1d(3,1,2) == 4)

    def test_to_2d(self):
        self.assertTrue(to_2d(5,0) == (0,0))
        self.assertTrue(to_2d(3,4) == (1,1))

    def test_solve(self):
        self.assertTrue(solve(self.d).board == self.e)    


if __name__ == '__main__':
    test = False
    if test:
        suite = unittest.TestLoader().loadTestsFromTestCase(TestSequenceFunctions)
        unittest.TextTestRunner(verbosity=2).run(suite)
    else:
        main()

File: Go/puzzle8.go
// Solve the 8 puzzle using the A* algorithm with heuristic manhattan distance + 
// the number of moves to reach the board as its negative priority.
// Uses the priority queue from http://golang.org/pkg/container/heap/ with modifications
// to resize the underlying array as needed. Assumes the puzzle has a solution

package main

import (
	"bufio"
	"bytes"
	"container/heap"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

//************************************************************************
// representation of the puzzle board
//************************************************************************

type Board struct {
	n         int     // dimension of board
	board     [][]int // the board in an nxn slice with 0 representing the blank tile
	manhattan int     // manhattan distance of board to solution board
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func newBoard(tiles [][]int) (b *Board) {
	// create a board form an nxn slice, and set its manhattan distance	
	b = new(Board)
	b.n = len(tiles[0][:])
	b.board = make([][]int, b.n)
	for i := 0; i < b.n; i++ {
		b.board[i] = make([]int, b.n)
		for j := 0; j < b.n; j++ {
			tileNum := tiles[i][j]
			b.board[i][j] = tileNum

			// calculate the row and col of the tile in the solution board 
			// and passing the board its manhattan distance
			goalrow := (tileNum - 1) / b.n
			goalcol := (tileNum - 1) % b.n
			if b.board[i][j] != 0 {
				b.manhattan += abs(i-goalrow) + abs(j-goalcol)
			}
		}
	}
	return b
}

func equals(x, y *Board) bool {
	// two boards are equal of all of the tiles are in the same place	
	if x.n != y.n {
		return false
	}
	for i := 0; i < x.n; i++ {
		for j := 0; j < x.n; j++ {
			if x.board[i][j] == y.board[i][j] {
				continue
			} else {
				return false
			}
		}
	}
	return true
}

func (b Board) String() string {
	// Board implements Stringer interface
	n := len(b.board[0][:])
	var buffer bytes.Buffer
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			s := fmt.Sprintf("%2d ", b.board[i][j])
			buffer.WriteString(s)
		}
		buffer.WriteString("\n")
	}
	return buffer.String()
}

func (b Board) neighbors() (boards []*Board) {
	// Calculate the possible resulting board after a legal move
	// return a slice with the 2 to 4 resulting boards
	N := b.n
	var row, col int
	tiles := make([][]int, N)
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			if b.board[i][j] == 0 {
				row, col = i, j
			}
		}
	}
	north := []int{-1, 1, 0, 0} // move 0 (blanck) tile up or down
	east := []int{0, 0, 1, -1}  // move 0 tile left or right
	for i := 0; i < 4; i++ {
		r := row + north[i]
		c := col + east[i]
		// if (r,c) is on the board (i.e. a legal move) create a board and add it
		// to the slice boards
		if !(r < 0 || r >= N || c < 0 || c >= N) {
			for m := 0; m < N; m++ {
				tiles[m] = make([]int, N)
				for n := 0; n < N; n++ {
					tiles[m][n] = b.board[m][n]
				}
			}
			tiles[row][col] = b.board[r][c]
			tiles[r][c] = 0
			nb := newBoard(tiles)
			if nb != nil {
				boards = append(boards, nb)
			}
		}
	}
	return boards
}

//************************************************************************
// Priority Queue contains elements of type *Node
//************************************************************************

type Node struct {
	board    *Board
	m        int   // moves made to arrive and this board
	priority int   // manhattan distance of board + m, smaller is better
	previous *Node // a link to previous node, used to output all boards in solution
	index    int   // the index of the item in the heap
}

type PriorityQueue []*Node // implements heap.Interface and holds Nodes

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	// We want Pop to give us the lowest, not highest, priority so we use less than here.
	// In other words we want a min-queue
	return pq[i].priority < pq[j].priority
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	// Push and Pop use pointer receivers because they modify the slice's length,
	// not just its contents. If the queue reaches capacity we copy the items into
	// a new slice with double the capcity.
	qLen := len(*pq)
	if qLen == cap(*pq) {
		c := make([]*Node, qLen, qLen*2)
		copy(c, *pq)
		*pq = c
	}
	*pq = (*pq)[0 : qLen+1]
	node := x.(*Node)
	node.index = qLen
	(*pq)[qLen] = node
}

func (pq *PriorityQueue) Pop() interface{} {
	// 	When queue size reaches 1/4 capacity, copy it to a new queue with 1/2 the capacity
	qLen := len(*pq)
	node := (*pq)[qLen-1]
	node.index = -1 // for safety
	*pq = (*pq)[0 : qLen-1]
	if len(*pq) < cap(*pq)/4 {
		c := make([]*Node, len(*pq)/2)
		copy(c, *pq)
		*pq = c
	}
	return node
}

//************************************************************************
// Solve the puzzle
//************************************************************************

func NewNode(b *Board, m int, previous *Node) (node *Node) {
	node = new(Node)
	node.board = b
	node.m = m
	node.previous = previous
	node.priority = b.manhattan + m
	return node
}

func Solve(initial *Board) *Node {
	initialNode := NewNode(initial, 0, nil)
	frontier := make(PriorityQueue, 0, 1)
	heap.Push(&frontier, initialNode)
	var node *Node
	for {
		node = heap.Pop(&frontier).(*Node)
		if node.board.manhattan == 0 {
			return node
		}
		for _, b := range node.board.neighbors() {
			if node.previous == nil || !equals(b, node.previous.board) {
				heap.Push(&frontier, NewNode(b, node.m+1, node))
			}
		}
	}
	return node
}

func ReadBoard() ([][]int) {
	f, err := os.Open(os.Args[1]) // command line argument is path to input file
	if err != nil {
		fmt.Println(err)
		return nil
	}
	defer f.Close()
	r := bufio.NewReader(f)
	line, err := r.ReadString('\n')
	s := strings.Trim(line, "\n")
	n, _ := strconv.Atoi(s)
	var tiles [][]int
	for i := 0; i < n; i++ {
		var row []int
		line, err = r.ReadString('\n')
		u := strings.Replace(line, "  ", " ", -1)
		u = strings.TrimSpace(u)
		t := strings.Split(u, " ")
		for _, w := range t {
			num, _ := strconv.Atoi(w)
			row = append(row, num)
		}
		tiles = append(tiles, row)
	}
	return tiles
}

func main() {
	// input file format is first row is the dimension, remaining rows are the values
	// of the board squares separated by one or two spaces.
	var bd = newBoard(ReadBoard())
	t0 := time.Now()
	node := Solve(bd)
	t1 := time.Now()
	fmt.Println("Solved in ", node.m, "moves, ", t1.Sub(t0), "seconds.\n")
	// use defer to pring boards of solution in order -- initial board first.
	defer fmt.Println(node.board)
	for node.previous != nil {
		node = node.previous
		defer fmt.Println(node.board)
	}
}
